File: packages/sdk/pact-py/src/pact/invariants/manifest.py
from __future__ import annotations

from typing import Any

from ..errors import parse_json_text
from .json import canonicalize_json
from .signing import is_valid_public_key_hex, public_key_hex_matches, verify_utf8_message_ed25519


def parse_signed_manifest_json(input_text: str) -> dict[str, Any]:
    return parse_json_text(input_text)


def signed_manifest_body_canonical_json(signed_manifest: dict[str, Any]) -> str:
    return canonicalize_json(signed_manifest["manifest"])


def _validate_manifest_structure(manifest: dict[str, Any]) -> bool:
    if manifest.get("schema") != "pact.manifest.v1":
        return False
    tools = manifest.get("tools", [])
    if not tools:
        return False
    seen: set[str] = set()
    for tool in tools:
        name = tool.get("name")
        if not isinstance(name, str) or name in seen:
            return False
        seen.add(name)
    return True


def verify_signed_manifest(signed_manifest: dict[str, Any]) -> dict[str, Any]:
    embedded_public_key_valid = is_valid_public_key_hex(
        signed_manifest["manifest"]["public_key"]
    )
    return {
        "structure_valid": _validate_manifest_structure(signed_manifest["manifest"]),
        "signature_valid": verify_utf8_message_ed25519(
            signed_manifest_body_canonical_json(signed_manifest),
            signed_manifest["signer_key"],
            signed_manifest["signature"],
        ),
        "embedded_public_key_valid": embedded_public_key_valid,
        "embedded_public_key_matches_signer": embedded_public_key_valid
        and public_key_hex_matches(
            signed_manifest["manifest"]["public_key"],
            signed_manifest["signer_key"],
        ),
    }


def verify_signed_manifest_json(input_text: str) -> dict[str, Any]:
    return verify_signed_manifest(parse_signed_manifest_json(input_text))

File: packages/sdk/chio-go/invariants/manifest.go
package invariants

type ManifestVerification struct {
	EmbeddedPublicKeyMatchesSigner bool `json:"embedded_public_key_matches_signer"`
	EmbeddedPublicKeyValid         bool `json:"embedded_public_key_valid"`
	SignatureValid                 bool `json:"signature_valid"`
	StructureValid                 bool `json:"structure_valid"`
}

func ParseSignedManifestJSON(input string) (map[string]any, error) {
	value, err := ParseJSONText(input)
	if err != nil {
		return nil, err
	}
	signedManifest, ok := value.(map[string]any)
	if !ok {
		return nil, newInvariantError("json", "signed manifest must be a JSON object")
	}
	return signedManifest, nil
}

func SignedManifestBodyCanonicalJSON(signedManifest map[string]any) (string, error) {
	manifest, err := mapField(signedManifest, "manifest")
	if err != nil {
		return "", err
	}
	return CanonicalizeJSON(manifest)
}

func VerifySignedManifest(signedManifest map[string]any) (ManifestVerification, error) {
	manifest, err := mapField(signedManifest, "manifest")
	if err != nil {
		return ManifestVerification{}, err
	}
	signature, err := stringField(signedManifest, "signature")
	if err != nil {
		return ManifestVerification{}, err
	}
	signerKey, err := stringField(signedManifest, "signer_key")
	if err != nil {
		return ManifestVerification{}, err
	}
	body, err := SignedManifestBodyCanonicalJSON(signedManifest)
	if err != nil {
		return ManifestVerification{}, err
	}
	signatureValid, err := VerifyUTF8MessageEd25519(body, signerKey, signature)
	if err != nil {
		return ManifestVerification{}, err
	}
	embeddedPublicKey, err := stringField(manifest, "public_key")
	if err != nil {
		return ManifestVerification{}, err
	}
	embeddedPublicKeyValid := IsValidEd25519PublicKeyHex(embeddedPublicKey)
	return ManifestVerification{
		EmbeddedPublicKeyMatchesSigner: embeddedPublicKeyValid && PublicKeyHexMatches(embeddedPublicKey, signerKey),
		EmbeddedPublicKeyValid:         embeddedPublicKeyValid,
		SignatureValid:                 signatureValid,
		StructureValid:                 validateManifestStructure(manifest),
	}, nil
}

func VerifySignedManifestJSON(input string) (ManifestVerification, error) {
	signedManifest, err := ParseSignedManifestJSON(input)
	if err != nil {
		return ManifestVerification{}, err
	}
	return VerifySignedManifest(signedManifest)
}

func validateManifestStructure(manifest map[string]any) bool {
	schema, ok := manifest["schema"].(string)
	if !ok || schema != "chio.manifest.v1" {
		return false
	}
	tools, ok := manifest["tools"].([]any)
	if !ok || len(tools) == 0 {
		return false
	}
	seen := make(map[string]struct{}, len(tools))
	for _, entry := range tools {
		tool, ok := entry.(map[string]any)
		if !ok {
			return false
		}
		name, ok := tool["name"].(string)
		if !ok || name == "" {
			return false
		}
		if _, exists := seen[name]; exists {
			return false
		}
		seen[name] = struct{}{}
	}
	return true
}
